README.md
# phlib/hashstrategy
PHP hash strategy library including random, sorted, consistent

[![Code Checks](https://img.shields.io/github/workflow/status/phlib/hashstrategy/CodeChecks?logo=github)](https://github.com/phlib/hashstrategy/actions/workflows/code-checks.yml)
[![Codecov](https://img.shields.io/codecov/c/github/phlib/hashstrategy.svg?logo=codecov)](https://codecov.io/gh/phlib/hashstrategy)
[![Latest Stable Version](https://img.shields.io/packagist/v/phlib/hashstrategy.svg?logo=packagist)](https://packagist.org/packages/phlib/hashstrategy)
[![Total Downloads](https://img.shields.io/packagist/dt/phlib/hashstrategy.svg?logo=packagist)](https://packagist.org/packages/phlib/hashstrategy)
![Licence](https://img.shields.io/github/license/phlib/hashstrategy.svg)

## Usage

### Strategies

Values are added to an index with a specific weighting.
They are fetched according to the strategy and the hashing `$key` passed to `get()`.

- Consistent
  - The same value(s) will always be returned for the given `$key`.
  - Weighting is used to increase the likelihood of a value being returned.
- Ordered
  - Values are returned in order of decreasing weight.
  - Values with the same weight are returned in the order added.
- Rand
  - Values are picked randomly.
  - Weighting is used to increase the likelihood of a value being returned.

```php
$pool = new \Phlib\HashStrategy\Consistent();
$pool->add('one');
$pool->add('two');
$pool->add('three');
var_dump($pool->get('hello', 2));
```

### ConfigPool

Combines available hash strategies with a set of configs to provide a direct way
to choose a config to use, for example choosing between a set of replicas.

```php
$config = [
     'server1' => ['hostname' => 'localhost', 'port' => 11211],
     'server2' => ['hostname' => 'localhost', 'port' => 11212],
     'server3' => ['hostname' => 'localhost', 'port' => 11213],
];
$hashStrategy = new \Phlib\HashStrategy\Consistent();
$pool = new \Phlib\HashStrategy\ConfigPool($config, $hashStrategy);
var_dump($pool->getManyConfigs('some key', 2));
```

## License

This package is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [2.0.0] - 2021-10-31
### Added
- Add support for PHP v7.4 & v8.0 .
- Type declarations have been added to all method parameters and return types
  where possible.
- Constant `Consistent::HASH_AVAILABLE` to make it easier to extend the class to
  add alternate hash methods.
### Changed
- **BC break**: Rename `Config` to `ConfigPool` for a more accurate description.
- **BC break**: Reduce visibility of internal methods and properties. These
  members are not part of the public API. No impact to standard use of this
  package. If an implementation has a use case which needs to override these
  members, please submit a pull request explaining the change.
### Removed
- **BC break**: Removed support for PHP versions <= v7.3 as they are no longer
  [actively supported](https://php.net/supported-versions.php) by the PHP project.

## [1.0.0] - 2015-02-22
- Initial release.


src/Config.php
<?php

declare(strict_types=1);

namespace Phlib\HashStrategy;

/**
 * Pool Config
 *
 * Used for hashing a pool of configs
 *
 * === Example ===
 * $config = array(
 *      'server1' => array('hostname' => 'localhost', 'port' => 11211),
 *      'server2' => array('hostname' => 'localhost', 'port' => 11212),
 *      'server3' => array('hostname' => 'localhost', 'port' => 11213),
 * );
 * $pool = new Phlib\HashStrategy\Config($config);
 * var_dump($pool->getConfigList('some key', 2));
 *
 * Class Config
 * @package Phlib\HashStrategy
 */
class Config
{
    private array $configList;

    private array $calculatedConfig = [];

    private HashStrategyInterface $hashStrategy;

    public function __construct(array $configList, HashStrategyInterface $hashStrategy = null)
    {
        // store the config array for later retrieval
        $this->configList = $configList;

        if (!$hashStrategy) {
            // no hash strategy was provided use a default
            $hashStrategy = new Ordered();
        }

        // setup the hashing
        $this->setHashStrategy($hashStrategy);
    }

    public function setHashStrategy(HashStrategyInterface $hashStrategy): self
    {
        // loop the config adding the key as a node
        foreach ($this->configList as $key => $value) {
            $hashStrategy->add(
                (string)$key,
                $value['weight'] ?? 1
            );
        }

        $this->hashStrategy = $hashStrategy;

        return $this;
    }

    public function getManyConfigs(string $key, int $count = 1): array
    {
        // find a calculated config list
        if (!array_key_exists("{$key}.{$count}", $this->calculatedConfig)) {
            // check we aren't storing too many calculated configs
            if (count($this->calculatedConfig) >= 100) {
                // remove the fist in the list, should be the oldest
                array_shift($this->calculatedConfig);
            }

            // get a list of config keys using the count and key provided
            $configList = [];
            foreach ($this->hashStrategy->get($key, $count) as $index) {
                // append the config values to the config list
                $configList[] = $this->configList[$index];
            }

            // store for later, a little config cache
            $this->calculatedConfig["{$key}.{$count}"] = $configList;
        }

        return $this->calculatedConfig["{$key}.{$count}"];
    }

    public function getConfig(string $key): array
    {
        // return the first matching config key
        $index = $this->hashStrategy->get($key, 1);

        return $this->configList[$index[0]];
    }

    public function getConfigList(): array
    {
        return $this->configList;
    }
}
